core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3578,12 +3578,13 @@ class KafkaApis(val requestChannel: RequestChannel,
           case Left(topLevelError) =>
             UpdateFeaturesResponse.createWithErrors(
               topLevelError,
-              Collections.emptyMap(),
+              Collections.emptySet(),
               throttleTimeMs)
           case Right(featureUpdateErrors) =>
+            // This is response is not correct, but since this is ZK specific code it will be removed in 4.0
             UpdateFeaturesResponse.createWithErrors(
               ApiError.NONE,
-              featureUpdateErrors.asJava,
+              featureUpdateErrors.asJava.keySet(),
```

Comment:
This is also used for None errors in the KafkaAdminClient test `testUpdateFeaturesHandleNotControllerException`. Maybe it is not totally necessary though. The other thing that is a little interesting about that test is that it doesn't serialize the result -- so it will always look like the v1 response.